### PR TITLE
add kcov 

### PR DESCRIPTION
**Pull request by davidtazy**
Specify library name and version:  **kcov/38**

- [ x] I've read the [guidelines](https://github.com/conan-io/conan-center-index/wiki#how-to-submit-a-pull-request) for contributing.
- [x ] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [ x] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [ x] I've tried at least one configuration locally with the
      [conan-center hook](https://github.com/conan-io/hooks.git) activated.

#1511

### CODE REVIEW

File: recipes/kcov/all/conanfile.py [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+        if self.settings.compiler == "Visual Studio":
+            raise ConanInvalidConfiguration(
+                  "kcov can not be built by Visual Studio.")
+
+    def source(self):
+        tools.get(**self.conan_data["sources"][self.version])
+        extracted_dir = self.name + "-" + self.version
+        os.rename(extracted_dir, self._source_subfolder)
+
+    def system_requirements(self):
```
Comment (uilianries) [inline recipes/kcov/all/conanfile.py:36]: System requirements should not be used in CCI.

File: recipes/kcov/all/test_package/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,6 @@
+cmake_minimum_required(VERSION 2.8.12)
```
Comment (uilianries) [inline recipes/kcov/all/test_package/CMakeLists.txt:1]: Use the same "template" than used for other recipes. For instance: https://github.com/conan-io/conan-center-index/tree/master/recipes/cereal/all/test_package

File: recipes/kcov/all/conanfile.py (hunk 1 above)
Comment (davidtazy) [reply recipes/kcov/all/conanfile.py:36]: @uilianries should i use it like backward-cpp does? warn if packages are note installed?
https://github.com/conan-io/conan-center-index/blob/37dda8777ea3a4cb15467f78acb9f8799d8e1092/recipes/backward-cpp/all/conanfile.py#L55-L81

Comment (uilianries) [reply recipes/kcov/all/conanfile.py:36]: Yes, it will be much better

**Comment by uilianries** [comment]
```
kcov/38: ERROR: while executing system_requirements(): kcov requires libdw-dev.
ERROR: Error in system requirements
```

Hmm, seems like we need to install the system requirements ....

**Comment by davidtazy** [comment]
Indeed

**Comment by uilianries** [comment]
@danimtb Can we inject it before running the build? Otherwise, we won't be able to make this package available.

**Comment by danimtb** [comment]
maybe the system requirements here could be just requires if those can be packages. I that possible @uilianries?

**Comment by uilianries** [comment]
@danimtb do you mean a proxy package? Yes, I think so.

@davidtazy I see this recipe requires libdw or libelf or binutils. What's the required library, libdw? I would say you could create proxy package for libdw, or even a regular package if possible. There is a good example here: https://github.com/conan-io/conan-center-index/blob/5480ada7d01cd8b3100704dd76c8571262069b7b/recipes/xorg/all/conanfile.py#L37

**Comment by davidtazy** [comment]
@uilianries , @danimtb , 
I'm a little confused because I copied the system_requirement method  from **accepted** backward-cpp recipe.

**Comment by SSE4** [comment]
right now, we try to avoid system requirements if possible. e.g. for `elfutils`, I guess we already have a package. we can package other dependencies first as well.

**Comment by uilianries** [comment]
We don't have elfutils in CCI yet, but there is a version in BIncrafters which can be ported: https://github.com/bincrafters/conan-elfutils

**Comment by uilianries** [comment]
Finally we have elfutils: https://github.com/conan-io/conan-center-index/pull/1768

@davidtazy please, update your recipe.

**Review by danimtb — APPROVED** [review]

File: recipes/kcov/all/conanfile.py [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+        self._patch_sources()
+        cmake = CMake(self)
+        cmake.configure(build_folder=self._build_subfolder)
+        cmake.build()
+
+    def package(self):
+        cmake = CMake(self)
+        cmake.configure(build_folder=self._build_subfolder)
+        cmake.install()
+        licdir = os.path.join(self.package_folder, "share", "doc", "kcov")
```
Comment (uilianries) [inline recipes/kcov/all/conanfile.py:51]:
```suggestion
        tools.rmdir(os.path.join(self.package_folder, "share"))
```

File: recipes/kcov/all/conanfile.py [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+        move(licdir, os.path.join(self.package_folder, "licenses"))
```
Comment (uilianries) [inline recipes/kcov/all/conanfile.py:52]:
```suggestion
        self.copy("COPYING*", dst="licenses", src=self._source_subfolder)
```

Changelog is not a license file. Copy directly

File: recipes/kcov/all/conanfile.py [hunk 5]
```diff
@@ -0,0 +1,59 @@
+import os
+from shutil import rmtree, move
```
Comment (uilianries) [inline recipes/kcov/all/conanfile.py:2]:
```suggestion
```

Not required

File: recipes/kcov/all/test_package/conanfile.py [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+        cmake.build()
+
+    def test(self):
+        if not tools.cross_building(self.settings):
+            bindir = os.path.join("bin", "test_package")
+            args = ["--exclude-pattern=/usr/",
+                    "--include-path={}".format(self.source_folder),
+                    "cov",
+                    bindir]
+            self.run("kcov {}".format(" ".join(args)), run_environment=True)
```
Comment (uilianries) [inline recipes/kcov/all/test_package/conanfile.py:23]:
```suggestion
    def test(self):
        if not tools.cross_building(self.settings):
            self.run("kcov --version", run_environment=True)
```

You can simply run --version subcommand to check the generated binary. The idea for test package is checking the generated artifact. So you can remove the test_package.cpp and CMake file.

File: recipes/kcov/all/conanfile.py [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+        cmake.configure(build_folder=self._build_subfolder)
+        cmake.build()
+
+    def package(self):
+        cmake = CMake(self)
+        cmake.configure(build_folder=self._build_subfolder)
+        cmake.install()
+        tools.rmdir(os.path.join(self.package_folder, "share"))
+        self.copy("COPYING*", dst="licenses", src=self._source_subfolder)
+        tools.rmdir(os.path.join(self.package_folder, "share"))
```
Comment (uilianries) [inline recipes/kcov/all/conanfile.py:52]:
```suggestion
```

duplicated

File: recipes/kcov/all/conanfile.py [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+                "kcov can not be built by Visual Studio.")
+
+    def source(self):
+        tools.get(**self.conan_data["sources"][self.version])
+        extracted_dir = self.name + "-" + self.version
+        os.rename(extracted_dir, self._source_subfolder)
+
+    def _patch_sources(self):
+        for patch in self.conan_data["patches"][self.version]:
+            tools.patch(**patch)
```
Comment (uilianries) [inline recipes/kcov/all/conanfile.py:38]:
```suggestion
            tools.patch(**patch)

    def _configure_cmake(self):
        if self._cmake is not None:
            return self._cmake
        self._cmake = CMake(self)
        self._cmake.configure(build_folder=self._build_subfolder)
        return self._cmake
```

File: recipes/kcov/all/conanfile.py [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+
+    def build(self):
+        self._patch_sources()
+        cmake = CMake(self)
+        cmake.configure(build_folder=self._build_subfolder)
```
Comment (uilianries) [inline recipes/kcov/all/conanfile.py:43]:
```suggestion
        cmake = self._configure_cmake()
```

File: recipes/kcov/all/conanfile.py [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+        cmake.build()
+
+    def package(self):
+        cmake = CMake(self)
+        cmake.configure(build_folder=self._build_subfolder)
```
Comment (uilianries) [inline recipes/kcov/all/conanfile.py:48]:
```suggestion
        cmake = self._configure_cmake()
```

File: recipes/kcov/all/test_package/conanfile.py [hunk 11]
```diff
@@ -0,0 +1,12 @@
+import os
+
+from conans import ConanFile, CMake, tools
```
Comment (uilianries) [inline recipes/kcov/all/test_package/conanfile.py:3]:
```suggestion
from conans import ConanFile, tools
```

File: recipes/kcov/all/test_package/conanfile.py [hunk 12]
```diff
@@ -0,0 +1,12 @@
...
+
+
+class KcovTestConan(ConanFile):
+    settings = "os", "compiler", "build_type", "arch"
+    generators = ["cmake"]
```
Comment (uilianries) [inline recipes/kcov/all/test_package/conanfile.py:8]:
```suggestion
```

File: recipes/kcov/all/conanfile.py [hunk 13]
```diff
@@ -0,0 +1,58 @@
...
+                "libiberty/9.1.0",
+                "libcurl/7.64.1",
+                "elfutils/0.180"]
+    generators = "cmake"
+
+    _source_subfolder = "source_subfolder"
+    _build_subfolder = "build_subfolder"
+
+    def configure(self):
+        if self.settings.compiler == "Visual Studio":
```
Comment (uilianries) [inline recipes/kcov/all/conanfile.py:27]: Can you build using mingw on Windows?

File: recipes/kcov/all/conanfile.py [hunk 14]
```diff
@@ -0,0 +1,58 @@
...
+            special compilation options"
+    topics = ("coverage", "linux", "debug")
+    settings = "os", "compiler", "build_type", "arch"
+    exports_sources = "CMakeLists.txt", "patches/**"
+    requires = ["zlib/1.2.11",
+                "libiberty/9.1.0",
+                "libcurl/7.64.1",
+                "elfutils/0.180"]
+    generators = "cmake"
+
```
Comment (uilianries) [inline recipes/kcov/all/conanfile.py:22]:
```suggestion
    _cmake = None
```

File: recipes/kcov/all/conanfile.py (hunk 13 above)
Comment (davidtazy) [reply recipes/kcov/all/conanfile.py:27]: I'm not the author, i'll ask him.
first line of readme.md  says : _"Kcov is a FreeBSD/Linux/OSX code coverage tester"_  .

Comment (uilianries) [reply recipes/kcov/all/conanfile.py:27]: Thanks, so can drop Windows if it is the case.

Comment (davidtazy) [reply recipes/kcov/all/conanfile.py:27]: ok i do it

**Review by SSE4 — APPROVED** [review]

**Review by danimtb — APPROVED** [review]

**Pull request merged**